### PR TITLE
[MI-2254] Added config setting to enable/disable account auto-connect

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
1. Added a boolean type field in system console.
2. Added a function to unregister current slash command on configuration change and register new command

### CODE REVIEW

File: server/plugin/plugin.go [hunk 1]
```diff
@@ -178,6 +180,16 @@ func (p *Plugin) OnConfigurationChange() (err error) {
 		return errors.New("status sync job interval should be greater than 0")
 	}
 
+	err = command.UnRegister(pluginAPIClient)
```
Comment (Kshitij-Katiyar) [inline server/plugin/plugin.go:183]:
```suggestion
	if err = command.UnRegister(pluginAPIClient); err !=nil {
```

File: server/plugin/plugin.go [hunk 2]
```diff
@@ -178,6 +180,16 @@ func (p *Plugin) OnConfigurationChange() (err error) {
...
+	if err != nil {
+		return errors.Wrap(err, "failed to unregister command")
+	}
+
+	err = command.Register(pluginAPIClient, stored.AutoConnectUsers)
```
Comment (Kshitij-Katiyar) [inline server/plugin/plugin.go:188]: same

File: server/plugin/plugin.go [hunk 3]
```diff
@@ -178,6 +180,16 @@ func (p *Plugin) OnConfigurationChange() (err error) {
...
+	if err != nil {
+		return errors.Wrap(err, "failed to register command")
```
Comment (Kshitij-Katiyar) [inline server/plugin/plugin.go:190]:
```suggestion
		return errors.Wrap(err, "failed to register slash command")
```

File: server/plugin/plugin.go [hunk 4]
```diff
@@ -370,16 +382,24 @@ func (p *Plugin) initEnv(e *Env, client *pluginapiclient.Client, pluginURL strin
 }
 
 func (p *Plugin) UserHasLoggedIn(c *plugin.Context, user *model.User) {
-	// Auto-connect the user to ms-calendar after he logs in
-	m := mscalendar.New(p.getEnv().Env, "")
-	_, err := m.GetRemoteUser(user.Id)
-	if err == nil {
-		// If user is already connected do nothing
+	stored := config.StoredConfig{}
+	err := p.API.LoadPluginConfiguration(&stored)
```
Comment (Kshitij-Katiyar) [inline server/plugin/plugin.go:386]: same

File: server/plugin/plugin.go [hunk 5]
```diff
@@ -370,16 +382,24 @@ func (p *Plugin) initEnv(e *Env, client *pluginapiclient.Client, pluginURL strin
...
-		p.API.LogError(fmt.Sprintf("error occurred while connecting user to mscalendar with email: %s. Error: %s", user.Email, err.Error()))
+	if stored.AutoConnectUsers {
+		// Auto-connect the user to ms-calendar after he logs in
+		m := mscalendar.New(p.getEnv().Env, "")
+		if _, err = m.GetRemoteUser(user.Id); err == nil {
+			// If user is already connected do nothing
+			return
+		}
+
+		err = m.CompleteOAuth2(user.Id)
```
Comment (Kshitij-Katiyar) [inline server/plugin/plugin.go:400]: here as well

**Review by Kshitij-Katiyar — APPROVED** [review]

File: plugin.json [hunk 6]
```diff
@@ -78,6 +78,13 @@
                 "display_name": "Status Sync Job Interval",
                 "type": "number",
                 "help_text": "Time in minutes after which the status sync job re-runs."
+            },
+            {
+                "key": "AutoConnectUsers",
+                "display_name": "Auto-Connect Users:",
+                "type": "bool",
+                "help_text": "When set to 'true' all the users on Mattermost are automatically connected to exchange-calender via ews-proxy-server.",
```
Comment (manojmalik20) [inline plugin.json:86]:
```suggestion
                "help_text": "When set to 'true', all the users on Mattermost are automatically connected to Exchange via EWS proxy server.",
```

File: server/command/command.go [hunk 7]
```diff
@@ -34,7 +34,7 @@ func getNotConnectedText() string {
 
 type handleFunc func(parameters ...string) (string, bool, error)
 
-var cmds = []*model.AutocompleteData{
+var commandsWhenAutoConnectIsFalse = []*model.AutocompleteData{
```
Comment (manojmalik20) [inline server/command/command.go:37]:
```suggestion
var commandsWhenAutoConnectDisabled = []*model.AutocompleteData{
```

File: server/command/command.go [hunk 8]
```diff
@@ -46,8 +46,26 @@ var cmds = []*model.AutocompleteData{
 	model.NewAutocompleteData("help", "", "Read help text for the commands"),
 }
 
+var commandsWhenAutoConnectIsTrue = []*model.AutocompleteData{
```
Comment (manojmalik20) [inline server/command/command.go:49]:
```suggestion
var commandsWhenAutoConnectEnabled = []*model.AutocompleteData{
```

File: server/command/help.go [hunk 9]
```diff
@@ -11,7 +11,10 @@ import (
 
 func (c *Command) help(parameters ...string) (string, bool, error) {
 	resp := ""
-	for _, cmd := range cmds {
+	for _, cmd := range commandsWhenAutoConnectIsFalse {
+		if c.Config.AutoConnectUsers && cmd.Trigger == "connect" {
```
Comment (manojmalik20) [inline server/command/help.go:15]: Is there a constant made for `connect`? If yes, you should use that.

File: server/plugin/plugin.go [hunk 10]
```diff
@@ -178,6 +180,14 @@ func (p *Plugin) OnConfigurationChange() (err error) {
 		return errors.New("status sync job interval should be greater than 0")
 	}
 
+	if err = command.UnRegister(pluginAPIClient); err != nil {
+		return errors.Wrap(err, "failed to unregister slash command")
+	}
+
+	if err = command.Register(pluginAPIClient, stored.AutoConnectUsers); err != nil {
+		return errors.Wrap(err, "failed to register slash command")
+	}
```
Comment (manojmalik20) [inline server/plugin/plugin.go:189]: These two blocks should be executed only if `previousConfig.AutoConnectUsers != config.AutoConnectUsers`.

File: server/plugin/plugin.go [hunk 11]
```diff
@@ -370,16 +380,22 @@ func (p *Plugin) initEnv(e *Env, client *pluginapiclient.Client, pluginURL strin
...
 
 func (p *Plugin) UserHasLoggedIn(c *plugin.Context, user *model.User) {
-	// Auto-connect the user to ms-calendar after he logs in
-	m := mscalendar.New(p.getEnv().Env, "")
-	_, err := m.GetRemoteUser(user.Id)
-	if err == nil {
-		// If user is already connected do nothing
+	stored := config.StoredConfig{}
+	if err := p.API.LoadPluginConfiguration(&stored); err != nil {
+		p.API.LogError(fmt.Sprintf("failed to load plugin configuration. Error: %s", err.Error()))
```
Comment (manojmalik20) [inline server/plugin/plugin.go:385]: Instead of this, can't you get the config from `p.getEnv()`?

File: server/plugin/plugin.go [hunk 12]
```diff
@@ -370,16 +380,22 @@ func (p *Plugin) initEnv(e *Env, client *pluginapiclient.Client, pluginURL strin
...
+	if stored.AutoConnectUsers {
+		// Auto-connect the user to ms-calendar after he logs in
+		m := mscalendar.New(p.getEnv().Env, "")
+		if _, err := m.GetRemoteUser(user.Id); err == nil {
+			// If user is already connected do nothing
+			return
+		}
+
+		if err := m.CompleteOAuth2(user.Id); err != nil {
+			p.API.LogError(fmt.Sprintf("error occurred while connecting user to mscalendar with email: %s. Error: %s", user.Email, err.Error()))
```
Comment (manojmalik20) [inline server/plugin/plugin.go:398]:
```suggestion
			p.API.LogError("Error occurred while connecting user to mscalendar", "Email", user.Email,  "Error", err.Error())
```

File: server/command/help.go (hunk 9 above)
Comment (raghavaggarwal2308) [reply server/command/help.go:15]: No, there is no constant

File: server/command/command.go [hunk 13]
```diff
@@ -46,7 +46,7 @@ var commandsWhenAutoConnectIsFalse = []*model.AutocompleteData{
 	model.NewAutocompleteData("help", "", "Read help text for the commands"),
 }
 
-var commandsWhenAutoConnectIsTrue = []*model.AutocompleteData{
+var commandsWhenAutoConnectEnabled = []*model.AutocompleteData{
```
Comment (manojmalik20) [inline server/command/command.go:49]:
```suggestion
var commandsWhenAutoConnectIsEnabled = []*model.AutocompleteData{
```

**Review by manojmalik20 — APPROVED** [review]

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**